File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // get previous (new next)
        ListNode* prev = nullptr;
        // get curr
        ListNode* curr = head;

        while (curr){
            // save curr's next
            ListNode* temp = curr->next;
            // point curr at previous
            curr->next = prev;

            // update prev to be curr
            prev = curr;
            // update curr to be curr's next
            curr = temp;
        }

        return prev;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        while True:
            toAdd = [float('inf'), None]
            for i, l in enumerate(lists):
                if l and l.val < toAdd[0]:
                    toAdd = [l.val, i]
            
            if toAdd[1] == None:
                break
            
            # get node in list
            node = lists[toAdd[1]]

            # add to our result
            curr.next = node

            # advance list in lists
            lists[toAdd[1]] = node.next
            # remove tie
            node.next = None

            # advance our result
            curr = curr.next

        return dummy.next
            